File: leetcode/1315-sum-of-nodes-with-even-valued-grandparent.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumEvenGrandparent(self, root: TreeNode) -> int:
        def dfs(node: TreeNode, parent_even: bool, gradparent_even: bool):
            if not node:
                return 0

            curr = node.val if gradparent_even  else 0
            curr_even = node.val%2 == 0
            return curr + dfs(node.left, curr_even, parent_even) + dfs(node.right, curr_even, parent_even)
        return dfs(root, False, False)

File: src/1315-sum-of-nodes-with-even-valued-grandparent.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumEvenGrandparent = function (root) {
  let acc = 0;

  function helper(node, padre, abuelo) {
    //caso base
    if (!node) return;
    //trabajo con el nodo actual
    if (abuelo != null && abuelo % 2 === 0) {
      acc += node.val;
    }
    //recursion
    helper(node.left, node.val, padre);
    helper(node.right, node.val, padre);
  }
  helper(root, null, null);
  return acc;
};
